File: group 2/variant13/loyalty_service.js
"use strict";

class Request {
    constructor(purchaseAmount, pointsToSpend, gold, promotion) {
        this.purchaseAmount = purchaseAmount;
        this.pointsToSpend = pointsToSpend;
        this.gold = gold;
        this.promotion = promotion;
        Object.freeze(this);
    }
}

class Result {
    constructor(status, amountDue = 0, pointsSpent = 0, pointsEarned = 0, balance = 0) {
        this.status = status;
        this.amountDue = amountDue;
        this.pointsSpent = pointsSpent;
        this.pointsEarned = pointsEarned;
        this.balance = balance;
        Object.freeze(this);
    }
}

class MemberStatus {
    isActive() {
        throw new Error("Потрібно надати залежність: MemberStatus");
    }
}

class PointsBalance {
    currentBalance() {
        throw new Error("Потрібно надати залежність: PointsBalance");
    }
}

class LoyaltyRepository {
    save(request, result) {
        throw new Error("Потрібно надати залежність: LoyaltyRepository");
    }
}

function checkRange(value, minimum, maximum, field) {
    if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
        throw new RangeError("Недопустиме значення: " + field);
    }
}

function checkFlags(...values) {
    if (values.some(value => typeof value !== "boolean")) {
        throw new TypeError("Ознаки мають бути логічними значеннями");
    }
}

function readFlag(value, field) {
    if (typeof value !== "boolean") {
        throw new Error("Некоректні відомості: " + field);
    }
    return value;
}

function readCount(value, field) {
    if (!Number.isSafeInteger(value) || value < 0) {
        throw new Error("Некоректні відомості: " + field);
    }
    return value;
}

class LoyaltyService {
    constructor(members, points, operations) {
        if (members == null || points == null || operations == null) {
            throw new TypeError("Усі залежності мають бути задані");
        }
        this.members = members;
        this.points = points;
        this.operations = operations;
    }

    process(request) {
        if (!(request instanceof Request)) {
            throw new TypeError("Заявку не задано або її формат некоректний");
        }
        checkRange(request.purchaseAmount, 100, 1_000_000, "Сума покупки");
        checkRange(request.pointsToSpend, 0, 4_999, "Бали для списання");
        checkFlags(request.gold, request.promotion);
        if (!readFlag(this.members.isActive(), "Активність рахунку")) {
            return new Result("Рахунок неактивний");
        }
        const balance = readCount(this.points.currentBalance(), "Баланс балів");
        if (request.pointsToSpend > balance) {
            return new Result("Недостатньо балів");
        }
        if (request.pointsToSpend >= Math.floor(request.purchaseAmount / 200)) {
            return new Result("Перевищено частку оплати");
        }
        const amountDue = request.purchaseAmount - request.pointsToSpend * 100;
        let multiplier;
        if (request.gold && request.promotion) {
            multiplier = 4;
        } else if (request.gold || request.promotion) {
            multiplier = 2;
        } else {
            multiplier = 1;
        }
        const earned = Math.floor(amountDue / 10_000) * multiplier;
        const updatedBalance = readCount(balance - request.pointsToSpend + earned, "Новий баланс балів");
        const result = new Result("Виконано", amountDue, request.pointsToSpend, earned, updatedBalance);
        this.operations.save(request, result);
        return result;
    }
}

module.exports = { Request, Result, MemberStatus, PointsBalance, LoyaltyRepository, LoyaltyService };

File: group 4/variant13/LoyaltyService.java
public final class LoyaltyService {
    public record Request(int purchaseAmount, int pointsToSpend, boolean gold, boolean promotion) {}

    public record Result(String status, long amountDue, int pointsSpent, long pointsEarned, long balance) {}

    public interface MemberStatus {
        boolean isActive();
    }

    public interface PointsBalance {
        long currentBalance();
    }

    public interface LoyaltyRepository {
        void save(Request request, Result result);
    }

    private final MemberStatus members;
    private final PointsBalance points;
    private final LoyaltyRepository operations;

    public LoyaltyService(MemberStatus members, PointsBalance points, LoyaltyRepository operations) {
        this.members = java.util.Objects.requireNonNull(members, "Залежність не задано: members");
        this.points = java.util.Objects.requireNonNull(points, "Залежність не задано: points");
        this.operations = java.util.Objects.requireNonNull(operations, "Залежність не задано: operations");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.purchaseAmount(), 100, 1_000_000, "Сума покупки");
        checkRange(request.pointsToSpend(), 0, 4_999, "Бали для списання");
        if (!members.isActive()) {
            return rejected("Рахунок неактивний");
        }
        long balance = points.currentBalance();
        if (balance < 0) {
            throw new IllegalStateException("Від’ємний баланс балів");
        }
        if (request.pointsToSpend() > balance) {
            return rejected("Недостатньо балів");
        }
        if (request.pointsToSpend() >= request.purchaseAmount() / 200) {
            return rejected("Перевищено частку оплати");
        }
        long amountDue = request.purchaseAmount() - request.pointsToSpend() * 100L;
        int multiplier;
        if (request.gold() && request.promotion()) {
            multiplier = 4;
        } else if (request.gold() || request.promotion()) {
            multiplier = 2;
        } else {
            multiplier = 1;
        }
        long earned = amountDue / 10_000 * multiplier;
        long updatedBalance = Math.addExact(balance - request.pointsToSpend(), earned);
        Result result = new Result("Виконано", amountDue, request.pointsToSpend(), earned, updatedBalance);
        operations.save(request, result);
        return result;
    }

    private static Result rejected(String status) {
        return new Result(status, 0, 0, 0, 0);
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}
